newbank/server/NewBankClientHandler.java

```diff
@@ -76,46 +80,87 @@ public void processRequests(CustomerID id) throws IOException {
+        if (request.getCommandName().equals("LOGOUT")) return;
       }
     }
 
-    private NewBankCommandResponse dispatch(NewBankCommandParameter parameter) {
+    private static class DispatchResult {
+
+      public DispatchResult(INewBankCommand command, NewBankCommandResponse response) {
+        this.command = command;
+        this.response = response;
+      }
 
-      switch (parameter.getCommandName()) {
+      private INewBankCommand command;
+      private NewBankCommandResponse response;
+
+      public String format() {
+
+        if (response.getType() == null) return "";
+
+        switch (response.getType()) {
+          case HELP:
+          case INVALIDREQUEST:
+            return response.getDescription().isBlank()
+                ? getHelpInfo()
+                : response.getDescription()
+                    + System.lineSeparator()
+                    + System.lineSeparator()
+                    + getHelpInfo();
+          default:
+            return response.getDescription();
+        }
+      }
+
+      private String getHelpInfo() {
+        return (this.command != null)
+            ? this.command.getCommandName() + " " + this.command.getDescription()
+            : "Unrecognised command";
+      }
+    }
+
+    private DispatchResult dispatch(NewBankCommandRequest request) {
+
+      switch (request.getCommandName()) {
         case "LOGOUT":
-          return NewBankCommandResponse.succeeded(null,
-              "Log out successful. Goodbye " + parameter.getId().getKey());
+          return new DispatchResult(
+              null, createSucceeded("Log out successful. Goodbye " + request.getId().getKey()));
         case "COMMANDS":
         case "HELP":
-          return NewBankCommandResponse.succeeded(null, formatCommands());
+          return new DispatchResult(null, createSucceeded(formatCommands()));
         default:
-          return runCommand(parameter);
+          return runCommand(request);
       }
     }
-    
-    private NewBankCommandResponse runCommand(NewBankCommandParameter parameter) {
-      if (commands.containsKey(parameter.getCommandName())) {
+
+    private DispatchResult runCommand(NewBankCommandRequest request) {
+      if (commands.containsKey(request.getCommandName())) {
+
+        INewBankCommand command = commands.get(request.getCommandName());
 
         // check if user is requesting help
-        if (parameter.getCommandArgument().matches("\\s*-([hH?]|help|HELP)\\s*$")) {
-          return NewBankCommandResponse.help(commands.get(parameter.getCommandName()));
-        } else {
-          return commands.get(parameter.getCommandName()).run(parameter);
-        }
+        if (request.getCommandArgument().matches("\\s*-([hH?]|help|HELP)\\s*$"))
+          return new DispatchResult(command, NewBankCommandResponse.createHelp());
+
+        NewBankCommandResponse response = new NewBankCommandResponse();
+        response.setStream(in, out);

```

Comment:
Hi Shinichi, could these in/out stream arguments be passed to the response constructor?  Is this the only reason we're not able to return the response from the command.run call?

NewBankCommandResponse response = command.run(request, in, out);    seems to flow better?

Reply:
Hi Simon, Yes, I can set the streams in a constructor.

As for the way of returning the response, it was a kind of difficult design decision. Let me explain.

I think we should encapsulate the streams and hide them from the Command implementations. 

By allowing the Invoker to take exclusive control of the streams, 

1. The Invoker is able to format the output in a standardized way in a single point because all the output is proved to go though the Invoker and its collaborators like the request and response classes. In other words, the Invoker works as [hooking](https://en.wikipedia.org/wiki/Hooking) for all the output operations.

2. The test and command implementations are decoupled with the stream interface and its implementation, and the decoupling makes those implementations simple and testable. A Stream has innately complicated API because it's general purpose. For example, a stream requires us to surround its I/O operations with try/catch block because a stream is an interface which could be bound to error-prone hardware like a network. Streams also change its behavior slightly but critically based on the media that it is bound to. When it bound to a network, an input operation is blocked while a user is typing. But when it is bound to memory, an input operation return null immediately at the end of the stream. These mean that each command implementation that accesses streams would potentially need to have code for test mode and for production mode in its inside.

It's also a difficult decision to make about how we could route the in/out streams to commands. I chose a Request/Response model. Like ASP.net or other web application framework, they have a request object and a response object in its runtime context, and each of which encapsulates input/output streams. I was not sure if it's an intuitive model to all of the members though, I decided to follow the common design pattern.

Reply:
Hi Shinichi thanks for the explanation.  Now that you've put it that way I think the encapsulation gains are worth the trade-off, or at least I can't see a better way to get around it.